### PR TITLE
Check `toMillis` overflow in operators with `Duration`

### LINKED ISSUE #868: Time.kt uses Duration.toMillis in unsafe manner
Using `Duration.ofSeconds(Long.MAX_VALUE)` as the timeout argument for [`kotlinx.coroutines.withTimeout`](https://github.com/Kotlin/kotlinx.coroutines/blob/d92b0fa7e5e5b5c19665cc649e8372bbe9cdf6e7/integration/kotlinx-coroutines-jdk8/src/time/Time.kt) will cause an ArithmaticOverflowException

This is because, in pure java `Durations.ofSeconds(Long.MAX_VALUE).toMillis()` will cause an Arithmatic overflow.

I've got a fix that basically involves adding an extension function `toMillisOrMax` that returns `Long.MAX_VALUE` when `toMillis` would otherwise overflow.

Use case:

I'm a big fan of null objects, and the way I've structured things is such that the "default" timeout is `Duration.of(seconds = Long.MAX_VALUE, nanos = 999__999_999)` (which is effectively `Duration.MAX_VALUE`), such that our down-stack code need not consider the case where no timeout is specified. 

If you guys are interested in this fix I'll submit a pull request to that involves simply:

```kt
private fun Duration.toMillisOrMax(): Long = when(this.seconds) {
    in 0L .. (Long.MAX_VALUE / MILLIS_PER_SECOND) -> duration.toMillis()
    else -> Long.MAX_VALUE
}
```

as a function in `TimeKt`, invoked instead of `Duration.toMillis()`

### PR DESCRIPTION
**Pull request by fvasco**
Fix #868

### CODE REVIEW

File: integration/kotlinx-coroutines-jdk8/src/time/Time.kt [hunk 1]
```diff
@@ -6,28 +6,44 @@ package kotlinx.coroutines.time
...
 public suspend fun <T> withTimeoutOrNull(duration: Duration, block: suspend CoroutineScope.() -> T): T? =
-    kotlinx.coroutines.withTimeoutOrNull(duration.toMillis(), block)
\ No newline at end of file
+        kotlinx.coroutines.withTimeoutOrNull(duration.toMillisOrMax(), block)
+
+/**
+ * Try to convert a [Duration] to milliseconds, negative durations are translated to `0`,
+ * too large durations are capped to `Duration.ofMillis(Long.MAX_VALUE)`.
+ */
+private fun Duration.toMillisOrMax(): Long {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-jdk8/src/time/Time.kt]: Please add some tests that don't pass without this change

File: integration/kotlinx-coroutines-jdk8/src/time/Time.kt [hunk 2]
```diff
@@ -6,28 +6,44 @@ package kotlinx.coroutines.time
...
 import kotlinx.coroutines.selects.SelectBuilder
 import java.time.Duration
-import java.util.concurrent.TimeUnit
 
 /**
  * "java.time" adapter method for [kotlinx.coroutines.delay]
  */
-public suspend fun delay(duration: Duration) =
-    kotlinx.coroutines.delay(duration.toMillis())
+public suspend fun delay(duration: Duration) {
```
Comment (qwwdfsad) [inline integration/kotlinx-coroutines-jdk8/src/time/Time.kt]: Please do not change codestyle of the methods

**Comment by fvasco** [comment]
Hi @qwwdfsad 
I added some test, I you need some extra ones please notify me.

I have to confide you a problem, `gradle test` does not work on any my computers.
So I had tested all `TimeTest` methods manually, but unfortunately I does not know if they compile. Sorry for my miss.
Providing a battle tested docker image is a great help for me.

**Comment by fvasco** [comment]
Please wait,
I wish to add some more tests and handle sub millisecond duration for all operators.

**Comment by fvasco** [comment]
Hi @qwwdfsad 
I catch this opportunity to fix another issue, any positive duration less than 1ms is truncated to 0ms, so no delay happens and `withTimeout` crashes.
To fix this issue all strictly positive duration are coerced in `1..Long.MAX_VALUE` milliseconds.
What do you think?

**Comment by qwwdfsad** [comment]
I think this is a good idea, thank you.
I've pushed your branch, improved documentation and tests and will merge it manually as soon as tests are passed.

Thank you for your contribution!

**Comment by qwwdfsad** [comment]
Merged manually

**Pull request closed without merging**